optional database name option

alanrenmsft opened the pull request:
check whether database name option is supported for the following scenarios
1. query editor title
2. databases dropdown in query editor toolbar
3. object explorer connection node label

**BEFORE**
Oracle: 
![image](https://user-images.githubusercontent.com/13777222/139481791-342aaa93-4c4e-4ce6-ba08-a747bc18bc2d.png)

SQL Server:
![image](https://user-images.githubusercontent.com/13777222/139481844-e9900cf9-f7bf-4154-adb2-3b52f88e1c85.png)

Object Explorer:
![image](https://user-images.githubusercontent.com/13777222/139486127-d29b54fb-6348-47de-9750-b5a7749afc25.png)


**AFTER**

Oracle:
![image](https://user-images.githubusercontent.com/13777222/139481704-fcb36235-4fac-42ea-b0b9-a0a60990a055.png)

SQL Server
![image](https://user-images.githubusercontent.com/13777222/139481776-fbfd1164-52f2-4479-941f-8c59b38e5423.png)

Object Explorer:
![image](https://user-images.githubusercontent.com/13777222/139486006-bce3b8a5-fbbc-44d6-9f28-31c1bb2b960a.png)

alanrenmsft commented on src/sql/workbench/contrib/query/browser/queryEditor.ts:
```diff
@@ -267,7 +270,6 @@ export class QueryEditor extends EditorPane {
 
 	private setTaskbarContent(): void {
 		// Create HTML Elements for the taskbar
-		const separator = Taskbar.createTaskbarSeparator();
```
this is a HtmlElement object, when we need multiple separators we have to create multiple objects. please compare the differences between the SQL Server toolbars in the PR description

nahk-ivanov reviewed (approved):
LGTM, but what do I know :)

Charles-Gagnon commented on src/sql/platform/connection/common/providerConnectionInfo.ts:
```diff
@@ -160,9 +160,13 @@ export class ProviderConnectionInfo extends Disposable implements azdata.Connect
-		let userName = this.userName ? this.userName : 'Windows Authentication';
-		return this.serverName + ', ' + databaseName + ' (' + userName + ')';
+		let title = this.serverName;
+		if (this.serverCapabilities.connectionOptions.find(option => option.specialValueType === ConnectionOptionSpecialType.databaseName)) {
```
Comment here about handling providers that don't support a database name

Charles-Gagnon commented on src/sql/platform/connection/common/providerConnectionInfo.ts:
```diff
@@ -160,9 +160,13 @@ export class ProviderConnectionInfo extends Disposable implements azdata.Connect
+			const databaseName = this.databaseName ? this.databaseName : '<default>';
+			title += `, ${databaseName}`;
+		}
+		title += ` (${this.userName ? this.userName : this.authenticationType})`;
```
Can also shorten this to `this.userName || this.authenticationType`

Charles-Gagnon commented on src/sql/workbench/common/editor/query/queryEditorInput.ts:
```diff
@@ -229,11 +229,11 @@ export abstract class QueryEditorInput extends EditorInput implements IConnectab
+				if (profile.databaseName) {
+					title += `.${profile.databaseName}`;
 				}
+				title += ` (${profile.userName ? profile.userName : profile.authenticationType})`;
```
`profiler.userName || profile.authenticationType`

Charles-Gagnon commented:
@nahk-ivanov Are you planning on having support for switching databases for Oracle via the UI? From what I understand this can be done once a connection is established - just not something that you can do in the initial connection.

nahk-ivanov commented:
@Charles-Gagnon No, not anymore. Originally I was trying to retrofit Oracle into the existing ADS model, but with this change there is one less thing for us to worry about.

The thing is - you can change it by updating session state, but so are many other session options. Given that on reconnect everything will be reset anyway - I don't see a reason why we would expose one session option, but not the others. I think it is fair that we provide them query interface and they can modify sessions state by executing a query, similarly to existing Oracle tools.

Charles-Gagnon commented:
Sounds good. Having a session state in addition to the connection profile is definitely interesting, could help a lot (especially as we onboard more types of data sources)

nahk-ivanov commented:
Well, I'm kinda torn here. Yeah, we started this "database context" things, but maintaining the entire session state will be very hard (and may not even be possible for some database platforms). Given that we can't have it done completely - I'm questioning whether we should even be doing it at all (even partially). Maybe all it takes is to make sure it is clear to the user when the state will be reset, i.e. showing a message saying "You got disconnected, we will automatically reconnect now, but your session state is busted. Make sure to restore the state so that your queries work as expected.". I initially thought there might be a feature to "clone" the session, something like open another query editor for the same "connection" (where "connection" is actually a session). But seems like there is none, so we don't really have to be aware of the remote session state on the client side.

I get it that for SQL Server database name is kind of unique in a sense that you can select it during connection and change it after connection (unless connected to PaaS), but for most DB engines that's not the case and "query context" is no different from any other option you configure on a session. I know in Oracle a lot of people define "login triggers" that configure the session right after connection, so they don't have to manually execute gazillion of commands to set current culture, etc.

The only thing where I can remotely think of the need to know current "query context" is intellisense/autocomplete for unqualified identifiers. Not sure if we do this already or plan to do it in ADS though.

alanrenmsft commented:
fixed a couple issues I found during the more comprehensive testing.
1. on startup, when the provider is not set for the connections in the OE, null ref exception will be logged in the console (feature works fine though)
2. on change connection/switch query editors opened for different providers, an error will occur because the database name is undefined.

also refactored the code in the query editor and added appropriate comments

Charles-Gagnon commented on src/sql/workbench/contrib/query/browser/queryActions.ts:
```diff
@@ -791,6 +791,11 @@ export class ListDatabasesActionItem extends Disposable implements IActionViewIt
+		if (!databaseName) {
+			return;
+		}
 		this._isConnected = true;
```
Wouldn't we still want to set this?

alanrenmsft replied:
I don't think it matters, because it is not visible anyways. when we change the connection to SQL Server, the code path will be executed again and all the states will be correctly set.

Charles-Gagnon replied:
Ah, didn't see that this was only within the context of the list databases action. (Would have been clearer to just name it enabled, but whatever)
